File: wysiwyg/src/main/java/com/github/onecode369/wysiwyg/WYSIWYG.kt
package com.github.onecode369.wysiwyg

import android.R
import android.annotation.SuppressLint
import android.content.Context
import android.content.res.TypedArray
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.util.AttributeSet
import android.util.Log
import android.view.Gravity
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.net.URLEncoder
import java.util.*
import kotlin.collections.ArrayList


class WYSIWYG @SuppressLint("SetJavaScriptEnabled") constructor(
    context: Context,
    attrs: AttributeSet?,
    defStyleAttr: Int
) :
    WebView(context, attrs, defStyleAttr) {
    enum class Type {
        BOLD,
        ITALIC,
        SUBSCRIPT,
        SUPERSCRIPT,
        STRIKETHROUGH,
        UNDERLINE,
        H1, H2, H3, H4, H5, H6,
        ORDEREDLIST,
        UNORDEREDLIST,
        JUSTIFYCENTER,
        JUSTIFYFULL,
        JUSTUFYLEFT,
        JUSTIFYRIGHT
    }



    interface OnTextChangeListener {
        fun onTextChange(text: String?)
    }

    interface OnDecorationStateListener {
        fun onStateChangeListener(
            text: String?,
            types: List<Type>?
        )
    }

    interface AfterInitialLoadListener {
        fun onAfterInitialLoad(isReady: Boolean)
    }

    private var isReady = false
    private var mContents: String? = null
    private var mTextChangeListener: OnTextChangeListener? = null
    private var mDecorationStateListener: OnDecorationStateListener? = null
    private var mLoadListener: AfterInitialLoadListener? = null

    constructor(context: Context) : this(context, null) {}
    constructor(context: Context, attrs: AttributeSet?) : this(
        context,
        attrs,
        R.attr.webViewStyle
    ) {
    }

    protected fun createWebviewClient(): EditorWebViewClient {
        return EditorWebViewClient()
    }

    fun setOnTextChangeListener(listener: OnTextChangeListener?) {
        mTextChangeListener = listener
    }

    fun setOnDecorationChangeListener(listener: OnDecorationStateListener?) {
        mDecorationStateListener = listener
    }

    fun setOnInitialLoadListener(listener: AfterInitialLoadListener?) {
        mLoadListener = listener
    }

    private fun callback(text: String) {
        mContents = text.replaceFirst(CALLBACK_SCHEME.toRegex(), "")
        if (mTextChangeListener != null) {
            mTextChangeListener!!.onTextChange(mContents)
        }
    }

    private fun stateCheck(text: String) {
        val state =
            text.replaceFirst(STATE_SCHEME.toRegex(), "")
                .toUpperCase(Locale.ENGLISH)
        val types: MutableList<Type> = ArrayList()
        for (type in Type.values()) {
            if (TextUtils.indexOf(state, type.name) != -1) {
                types.add(type)
            }
        }
        if (mDecorationStateListener != null) {
            mDecorationStateListener!!.onStateChangeListener(state, types)
        }
    }

    private fun applyAttributes(context: Context, attrs: AttributeSet?) {
        val attrsArray = intArrayOf(
            R.attr.gravity
        )
        val ta: TypedArray = context.obtainStyledAttributes(attrs, attrsArray)
        val gravity = ta.getInt(0, View.NO_ID)
        when (gravity) {
            Gravity.LEFT -> exec("javascript:editor.setTextAlign(\"left\")")
            Gravity.RIGHT -> exec("javascript:editor.setTextAlign(\"right\")")
            Gravity.TOP -> exec("javascript:editor.setVerticalAlign(\"top\")")
            Gravity.BOTTOM -> exec("javascript:editor.setVerticalAlign(\"bottom\")")
            Gravity.CENTER_VERTICAL -> exec("javascript:editor.setVerticalAlign(\"middle\")")
            Gravity.CENTER_HORIZONTAL -> exec("javascript:editor.setTextAlign(\"center\")")
            Gravity.CENTER -> {
                exec("javascript:editor.setVerticalAlign(\"middle\")")
                exec("javascript:editor.setTextAlign(\"center\")")
            }
        }
        ta.recycle()
    }

    // No handling
    var html: String?
        get() = mContents
        set(contents) {
            var contents = contents
            if (contents == null) {
                contents = ""
            }
            try {
                exec(
                    "javascript:editor.setHtml('" + URLEncoder.encode(
                        contents,
                        "UTF-8"
                    ).toString() + "');"
                )
            } catch (e: UnsupportedEncodingException) { // No handling
            }
            mContents = contents
        }

    fun setEditorFontColor(color: Int) {
        val hex = convertHexColorString(color)
        exec("javascript:editor.setBaseTextColor('$hex');")
    }

    fun setEditorFontSize(px: Int) {
        exec("javascript:editor.setBaseFontSize('" + px + "px');")
    }

    override fun setPadding(left: Int, top: Int, right: Int, bottom: Int) {
        super.setPadding(left, top, right, bottom)
        exec(
            "javascript:editor.setPadding('" + left + "px', '" + top + "px', '" + right + "px', '" + bottom
                    + "px');"
        )
    }

    override fun setPaddingRelative(
        start: Int,
        top: Int,
        end: Int,
        bottom: Int
    ) { // still not support RTL.
        setPadding(start, top, end, bottom)
    }

    fun setEditorBackgroundColor(color: Int) {
        setBackgroundColor(color)
    }

    override fun setBackgroundColor(color: Int) {
        super.setBackgroundColor(color)
    }

    override fun setBackgroundResource(resid: Int) {
        val bitmap: Bitmap = Utils.decodeResource(context, resid)
        val base64: String = Utils.toBase64(bitmap)
        bitmap.recycle()
        exec("javascript:editor.setBackgroundImage('url(data:image/png;base64,$base64)');")
    }

    override fun setBackground(background: Drawable) {
        val bitmap: Bitmap = Utils.toBitmap(background)
        val base64: String = Utils.toBase64(bitmap)
        bitmap.recycle()
        exec("javascript:editor.setBackgroundImage('url(data:image/png;base64,$base64)');")
    }

    fun setBackground(url: String) {
        exec("javascript:editor.setBackgroundImage('url($url)');")
    }

    fun setEditorWidth(px: Int) {
        exec("javascript:editor.setWidth('" + px + "px');")
    }

    fun setEditorHeight(px: Int) {
        exec("javascript:editor.setHeight('" + px + "px');")
    }

    fun setPlaceholder(placeholder: String) {
        exec("javascript:editor.setPlaceholder('$placeholder');")
    }

    fun setInputEnabled(inputEnabled: Boolean) {
        exec("javascript:editor.setInputEnabled($inputEnabled)")
    }

    fun loadCSS(cssFile: String) {
        val jsCSSImport = "(function() {" +
                "    var head  = document.getElementsByTagName(\"head\")[0];" +
                "    var link  = document.createElement(\"link\");" +
                "    link.rel  = \"stylesheet\";" +
                "    link.type = \"text/css\";" +
                "    link.href = \"" + cssFile + "\";" +
                "    link.media = \"all\";" +
                "    head.appendChild(link);" +
                "}) ();"
        exec("javascript:$jsCSSImport")
    }

    fun undo() {
        exec("javascript:editor.undo();")
    }

    fun redo() {
        exec("javascript:editor.redo();")
    }

    fun setBold() {
        exec("javascript:editor.setBold();")
    }

    fun setItalic() {
        exec("javascript:editor.setItalic();")
    }

    fun setSubscript() {
        exec("javascript:editor.setSubscript();")
    }

    fun setSuperscript() {
        exec("javascript:editor.setSuperscript();")
    }

    fun setStrikeThrough() {
        exec("javascript:editor.setStrikeThrough();")
    }

    fun setUnderline() {
        exec("javascript:editor.setUnderline();")
    }

    fun setTextColor(color: Int) {
        exec("javascript:editor.prepareInsert();")
        val hex = convertHexColorString(color)
        exec("javascript:editor.setTextColor('$hex');")
    }

    fun setTextBackgroundColor(color: Int) {
        exec("javascript:editor.prepareInsert();")
        val hex = convertHexColorString(color)
        exec("javascript:editor.setTextBackgroundColor('$hex');")
    }

    fun setFontSize(fontSize: Int) {
        if (fontSize > 7 || fontSize < 1) {
            Log.e("WYSIWYG", "Font size should have a value between 1-7")
        }
        exec("javascript:editor.setFontSize('$fontSize');")
    }

    fun removeFormat() {
        exec("javascript:editor.removeFormat();")
    }

    fun setHeading(heading: Int) {
        exec("javascript:editor.setHeading('$heading');")
    }

    fun setCode() {
        exec("javascript:editor.setCode();")
    }

    fun setIndent() {
        exec("javascript:editor.setIndent();")
    }

    fun setOutdent() {
        exec("javascript:editor.setOutdent();")
    }

    fun setAlignLeft() {
        exec("javascript:editor.setJustifyLeft();")
    }

    fun setAlignCenter() {
        exec("javascript:editor.setJustifyCenter();")
    }

    fun setAlignJustifyFull() {
        exec("javascript:editor.setJustifyFull();")
    }

    fun setAlignRight() {
        exec("javascript:editor.setJustifyRight();")
    }

    fun setBlockquote() {
        exec("javascript:editor.setBlockquote();")
    }

    fun setBullets() {
        exec("javascript:editor.setBullets();")
    }

    fun setNumbers() {
        exec("javascript:editor.setNumbers();")
    }

    fun insertImage(url: String, alt: String) {
        exec("javascript:editor.prepareInsert();")
        exec("javascript:editor.insertImage('$url', '$alt');")
    }

    fun insertLink(href: String, title: String) {
        exec("javascript:editor.prepareInsert();")
        exec("javascript:editor.insertLink('$href', '$title');")
    }

    fun insertTodo() {
        exec("javascript:editor.prepareInsert();")
        exec("javascript:editor.setTodo('" + Utils.currentTime.toString() + "');")
    }

    fun focusEditor() {
        requestFocus()
        exec("javascript:editor.focus();")
    }

    fun clearFocusEditor() {
        exec("javascript:editor.blurFocus();")
    }

    private fun convertHexColorString(color: Int): String {
        return String.format("#%06X", 0xFFFFFF and color)
    }

    protected fun exec(trigger: String) {
        if (isReady) {
            load(trigger)
        } else {
            postDelayed({ exec(trigger) }, 100)
        }
    }

    private fun load(trigger: String) {
        evaluateJavascript(trigger, null)
    }

    protected inner class EditorWebViewClient : WebViewClient() {
        override fun onPageFinished(view: WebView, url: String) {
            isReady = url.equals(SETUP_HTML, ignoreCase = true)
            if (mLoadListener != null) {
                mLoadListener!!.onAfterInitialLoad(isReady)
            }
        }

        override fun shouldOverrideUrlLoading(
            view: WebView,
            url: String
        ): Boolean {
            val decode: String
            decode = try {
                URLDecoder.decode(url, "UTF-8")
            } catch (e: UnsupportedEncodingException) { // No handling
                return false
            }
            if (TextUtils.indexOf(url, CALLBACK_SCHEME) == 0) {
                callback(decode)
                return true
            } else if (TextUtils.indexOf(url, STATE_SCHEME) == 0) {
                stateCheck(decode)
                return true
            }
            return super.shouldOverrideUrlLoading(view, url)
        }
    }

    companion object {
        private const val SETUP_HTML = "file:///android_asset/editor.html"
        private const val CALLBACK_SCHEME = "re-callback://"
        private const val STATE_SCHEME = "re-state://"
    }


    init {
        isVerticalScrollBarEnabled = false
        isHorizontalScrollBarEnabled = false
        settings.javaScriptEnabled = true
        webChromeClient = WebChromeClient()
        webViewClient = createWebviewClient()

        loadUrl(SETUP_HTML)
        applyAttributes(context, attrs)
    }
}

File: wysiwyg/src/main/assets/wysiwyg.js
var editor = {};

editor.currentSelection = {
    "startContainer": 0,
    "startOffset": 0,
    "endContainer": 0,
    "endOffset": 0};

editor.editor = document.getElementById('editor');



document.addEventListener("selectionchange", function() { editor.backuprange(); });

// Initializations
editor.callback = function() {
    window.location.href = "re-callback://" + encodeURI(editor.getHtml());
}

editor.setHtml = function(contents) {
    editor.editor.innerHTML = decodeURIComponent(contents.replace(/\+/g, '%20'));
}

editor.getHtml = function() {
    return editor.editor.innerHTML;
}

editor.getText = function() {
    return editor.editor.innerText;
}

editor.setBaseTextColor = function(color) {
    editor.editor.style.color  = color;
}

editor.setBaseFontSize = function(size) {
    editor.editor.style.fontSize = size;
}

editor.setPadding = function(left, top, right, bottom) {
  editor.editor.style.paddingLeft = left;
  editor.editor.style.paddingTop = top;
  editor.editor.style.paddingRight = right;
  editor.editor.style.paddingBottom = bottom;
}

editor.setBackgroundColor = function(color) {
    document.body.style.backgroundColor = color;
}

editor.setBackgroundImage = function(image) {
    editor.editor.style.backgroundImage = image;
}

editor.setWidth = function(size) {
    editor.editor.style.minWidth = size;
}

editor.setHeight = function(size) {
    editor.editor.style.height = size;
}

editor.setTextAlign = function(align) {
    editor.editor.style.textAlign = align;
}

editor.setVerticalAlign = function(align) {
    editor.editor.style.verticalAlign = align;
}

editor.setPlaceholder = function(placeholder) {
    editor.editor.setAttribute("placeholder", placeholder);
}

editor.setInputEnabled = function(inputEnabled) {
    editor.editor.contentEditable = String(inputEnabled);
}

editor.undo = function() {
    document.execCommand('undo', false, null);
}

editor.redo = function() {
    document.execCommand('redo', false, null);
}

editor.setBold = function() {
    document.execCommand('bold', false, null);
}

editor.setItalic = function() {
    document.execCommand('italic', false, null);
}

editor.setSubscript = function() {
    if(window.getSelection().toString() != "")
    var html = '<sub>' + window.getSelection().toString() + '</sub> &nbsp;';
    editor.insertHTML(html);
}

editor.setSuperscript = function() {
    if(window.getSelection().toString() != "")
    var html = '<sup>' + window.getSelection().toString() + '</sup> &nbsp;';
    editor.insertHTML(html);
}

editor.setStrikeThrough = function() {
    document.execCommand('strikeThrough', false, null);
}

editor.setUnderline = function() {
    document.execCommand('underline', false, null);
}

editor.setBullets = function() {
    document.execCommand('insertUnorderedList', false, null);
}

editor.setNumbers = function() {
    document.execCommand('insertOrderedList', false, null);
}

editor.setTextColor = function(color) {
    editor.restorerange();
    document.execCommand("styleWithCSS", null, true);
    document.execCommand('foreColor', false, color);
    document.execCommand("styleWithCSS", null, false);
}

editor.setTextBackgroundColor = function(color) {
    editor.restorerange();
    document.execCommand("styleWithCSS", null, true);
    document.execCommand('hiliteColor', false, color);
    document.execCommand("styleWithCSS", null, false);
}

editor.setFontSize = function(fontSize){
    document.execCommand("fontSize", false, fontSize);
}

editor.setHeading = function(heading) {
    document.execCommand('formatBlock', false, '<h'+heading+'>');
}

editor.setIndent = function() {
    document.execCommand('indent', false, null);
}

editor.setOutdent = function() {
    document.execCommand('outdent', false, null);
}

editor.setJustifyLeft = function() {
    document.execCommand('justifyLeft', false, null);
}

editor.setJustifyCenter = function() {
    document.execCommand('justifyCenter', false, null);
}

editor.setJustifyRight = function() {
    document.execCommand('justifyRight', false, null);
}

editor.setJustifyFull = function() {
    document.execCommand('justifyFull', false, null);
}

editor.setBlockquote = function() {
    if(window.getSelection().toString() != "")
    var html = '<p style="color: #000 !important;background-color: #f1f1f1 !important;border-left: 6px solid #ccc !important;padding:10px">' + window.getSelection().toString() + '</p></br>';
    editor.insertHTML(html);
}

editor.setCode = function() {
    if(window.getSelection().toString() != ""){
        var code = window.getSelection().toString().replace(/</ig, '&lt;').replace(/>/ig, '&gt;');
        var html ='<div class="w3-panel w3-card w3-light-grey"><div class="w3-code htmlHigh notranslate">'+code+'</div></div></br>'
    }
    editor.insertHTML(html);
}

editor.insertImage = function(url, alt) {
    var html = '<img style="width:100%;max-width:400px" src="' + url + '" alt="' + alt + '" />';
    editor.insertHTML(html);
}

editor.insertLatex = function(latex) {
    var html = '<img src="https://private.codecogs.com/png.download?' + latex + '" alt="' + latex + '" />';
    editor.insertHTML(html);
}

editor.insertHTML = function(html) {
    editor.restorerange();
    document.execCommand('insertHTML', false, html);
}

editor.insertLink = function(url, title) {
    editor.restorerange();
    var sel = document.getSelection();
    if (sel.toString().length == 0) {
        document.execCommand("insertHTML",false,"<a href='"+url+"'>"+title+"</a>");
    } else if (sel.rangeCount) {
       var el = document.createElement("a");
       el.setAttribute("href", url);
       el.setAttribute("title", title);

       var range = sel.getRangeAt(0).cloneRange();
       range.surroundContents(el);
       sel.removeAllRanges();
       sel.addRange(range);
   }
    editor.callback();
}

editor.setTodo = function(text) {
    var html = '<input class="w3-checkbox" type="checkbox" name="'+ text +'" value="'+ text +'"/> &nbsp;';
    document.execCommand('insertHTML', false, html);
}

editor.setTodo = function(text) {
    var html = '<input class="w3-checkbox" type="checkbox" name="'+ text +'" value="'+ text +'"/> &nbsp;';
    document.execCommand('insertHTML', false, html);
}

editor.prepareInsert = function() {
    RE.backuprange();
}

editor.backuprange = function(){
    var selection = window.getSelection();
    if (selection.rangeCount > 0) {
      var range = selection.getRangeAt(0);
      editor.currentSelection = {
          "startContainer": range.startContainer,
          "startOffset": range.startOffset,
          "endContainer": range.endContainer,
          "endOffset": range.endOffset};
    }
}

editor.restorerange = function(){
    var selection = window.getSelection();
    selection.removeAllRanges();
    var range = document.createRange();
    range.setStart(editor.currentSelection.startContainer, editor.currentSelection.startOffset);
    range.setEnd(editor.currentSelection.endContainer, editor.currentSelection.endOffset);
    selection.addRange(range);
}

editor.enabledEditingItems = function(e) {
    var items = [];
    if (document.queryCommandState('bold')) {
        items.push('bold');
    }
    if (document.queryCommandState('italic')) {
        items.push('italic');
    }
    if (document.queryCommandState('subscript')) {
        items.push('subscript');
    }
    if (document.queryCommandState('superscript')) {
        items.push('superscript');
    }
    if (document.queryCommandState('strikeThrough')) {
        items.push('strikeThrough');
    }
    if (document.queryCommandState('underline')) {
        items.push('underline');
    }
    if (document.queryCommandState('insertOrderedList')) {
        items.push('orderedList');
    }
    if (document.queryCommandState('insertUnorderedList')) {
        items.push('unorderedList');
    }
    if (document.queryCommandState('justifyCenter')) {
        items.push('justifyCenter');
    }
    if (document.queryCommandState('justifyFull')) {
        items.push('justifyFull');
    }
    if (document.queryCommandState('justifyLeft')) {
        items.push('justifyLeft');
    }
    if (document.queryCommandState('justifyRight')) {
        items.push('justifyRight');
    }
    if (document.queryCommandState('insertHorizontalRule')) {
        items.push('horizontalRule');
    }
    var formatBlock = document.queryCommandValue('formatBlock');
    if (formatBlock.length > 0) {
        items.push(formatBlock);
    }

    window.location.href = "re-state://" + encodeURI(items.join(','));
}

editor.focus = function() {
    var range = document.createRange();
    range.selectNodeContents(editor.editor);
    range.collapse(false);
    var selection = window.getSelection();
    selection.removeAllRanges();
    selection.addRange(range);
    editor.editor.focus();
}

editor.blurFocus = function() {
    editor.editor.blur();
}

editor.removeFormat = function() {
    document.execCommand('removeFormat', false, null);
}

// Event Listeners
editor.editor.addEventListener("input", editor.callback);
editor.editor.addEventListener("keyup", function(e) {
    var KEY_LEFT = 37, KEY_RIGHT = 39;
    if (e.which == KEY_LEFT || e.which == KEY_RIGHT) {
        editor.enabledEditingItems(e);
    }
});
editor.editor.addEventListener("click", editor.enabledEditingItems);
